### ISSUE TITLE
Borrow features from Mona

### ISSUE THREAD

**Issue by wavebeem** [jneen/parsimmon#168]
I've been playing with https://github.com/zkat/mona and there are a few functions that would be nice to adapt into Parsimmon functions/methods.

## trim

This is an extremely common parsing operation. The only caveat is obviously it doesn't cover comments, but `wrap` can cover that use case.

```js
P.digits
  .trim(P.optWhitespace)
  .tryParse('     123  \n\n\r\t     ')
// => '123'
```

is equivalent to…

```js
P.optWhitespace
  .then(P.digits)
  .skip(P.optWhitespace)
  .tryParse('     123  \n\n\r\t     ')
// => '123'
```

## wrap

Pretty much every language needs something like this, and I think this is a lot less clunky than using `seqMap` or `.then` + `.skip`.

```js
P.digits
  .wrap(P.string('('), P.string(')'))
  .tryParse('(123)')
// '123'
```

Like `trim` but you pass two parsers for the before/after. Probably would often get chained off of `trim` though.

## join

This would definitely come in handy for the lower level bits of parsing, especially parsing identifiers and numbers.

```js
P.join(
  P.oneOf('+-').fallback(''),
  P.digits,
  P.string('.'),
  P.digits,
  P.join(
    P.oneOf('eE'),
    P.oneOf('+-').fallback(''),
    P.digits
  )
).tryParse('-3.14e+4')
// => '-3.14e+4'
```

Like `seq` but just concatenates all the values.

## range

`oneOf` is nice but this covers a subset of the `regex` parser that's definitely not represented elsewhere.

```js
P.range('a', 'z').tryParse('x')
// => x
```

Basically works like:

```js
P.test(c => c >= 'a' && c <= 'z')
```

To just do an inclusive character range check.

**Comment by wavebeem** [comment]
Question: Should `.trim()` have an optional argument for a different parser to use besides `P.optWhitespace`, or should it be mandatory?

I'm leaning towards mandatory so it's obvious that you can (and *should*) replace it to suit your needs.

**Comment by wavebeem** [comment]
I think I'll actually go with a `.join()` method over `P.text`:

## P.text

```js
P.text(
  P.oneOf('+-').fallback(''),
  P.digits,
  P.string('.'),
  P.digits,
  P.text(
    P.oneOf('eE'),
    P.oneOf('+-').fallback(''),
    P.digits
  )
).tryParse('-3.14e+4')
// => '-3.14e+4'
```

## parser.join()

```js
P.seq(
  P.oneOf('+-').fallback(''),
  P.digits,
  P.string('.'),
  P.digits,
  P.seq(
    P.oneOf('eE'),
    P.oneOf('+-').fallback(''),
    P.digits
  ).join().fallback('')
).join().tryParse('-3.14e+4')
// => '-3.14e+4'
```

**Comment by caballus** [comment]
mona's [sequence](https://github.com/zkat/mona#mona-sequence) is also very nice feature to have.
With that you can understand what is going on inside of your parser better. Put breakpoints etc.

Also it can be faster than seqMap, since it does not create intermediate arrays. But since it probably exits by throwing and catching exceptions and v8 doesn't like to optimize that, it can actually be slower. It is a subject to test I think.

**Comment by wavebeem** [comment]
Yeah, I imagine it's *probably* a good bit slower just because it involves exceptions. Would want to test. I'm currently thinking about a more flexible alternative to `seq`, so you might want to keep an eye out for when I open that ticket instead.

**Comment by caballus** [comment]
It can also be done with generators. But generators also comes with some performance overhead
